src/main/java/com/uber/cadence/internal/shadowing/ReplayWorkflowActivityImpl.java

```diff
@@ -0,0 +1,133 @@
+import com.uber.cadence.testing.WorkflowReplayer;
+import com.uber.m3.tally.Scope;
+import java.util.List;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ReplayWorkflowActivityImpl implements ReplayWorkflowActivity {
+
+  private static final Logger log = LoggerFactory.getLogger(ReplayWorkflowActivityImpl.class);
+
+  private final IWorkflowService serviceClient;
+  private final boolean failFast;
+  private final Scope metricsScope;
+  private final Class<?> workflowClass;
+  private final Class<?>[] workflowClasses;
+
+  public ReplayWorkflowActivityImpl(
+      IWorkflowService serviceClient,
+      boolean failFast,
+      Scope metricsScope,
+      Class<?> workflowClass,
+      Class<?>... workflowClasses) {
+    this.serviceClient = serviceClient;
+    this.failFast = failFast;
+    this.metricsScope = metricsScope;
+    this.workflowClass = workflowClass;
+    this.workflowClasses = workflowClasses;
+  }
+
+  @Override
+  public ReplayWorkflowActivityResponse replay(ReplayWorkflowActivityRequest request)
+      throws Throwable {
+    if (request == null) {
+      throw new NullPointerException("Replay activity request is null.");
+    }
+
+    String domain = request.getDomain();
+    int successCount = 0;
+    int failedCount = 0;
+    int skippedCount = 0;
+    List<WorkflowExecution> failedExecution = Lists.newArrayList();
+    for (WorkflowExecution execution : request.getExecutions()) {
+
+      WorkflowExecutionHistory workflowHistory;
+      try {
+        workflowHistory = getFullHistory(domain, execution);
+      } catch (Exception e) {
+        log.error(
+            "failed to get workflow history with domain: "
+                + request.getDomain()
+                + " "
+                + execution.toString(),
+            e);
+        skippedCount++;
+        continue;
+      }
+
+      try {
+        WorkflowReplayer.replayWorkflowExecution(workflowHistory, workflowClass, workflowClasses);
+      } catch (Exception e) {
+        log.error(
+            "failed to replay workflow history with domain: "
+                + request.getDomain()
+                + " "
+                + execution.toString(),
+            e);
+
+        // Count any error with keyword 'nondeterministic' as non deterministic error.
+        if (e.getCause().getMessage().contains("nondeterministic")) {
+          failedCount++;
+          failedExecution.add(execution);
+          this.metricsScope.counter(MetricsType.SHADOWING_NON_DETERMINISTIC_ERROR).inc(1);
+          if (this.failFast) {
+            return new ReplayWorkflowActivityResponse(
+                successCount, failedCount, skippedCount, failedExecution);
+          }
+        } else {
+          this.metricsScope.counter(MetricsType.SHADOWING_SKIPPED_COUNT).inc(1);
+          skippedCount++;
+        }
+        continue;
+      }
+      this.metricsScope.counter(MetricsType.SHADOWING_SUCCESS_COUNT).inc(1);
+      successCount++;
```

Comment:
shall we use heartbeat to record the current progress?

Reply:
Yes. Updated